R/plotstyle.R

```diff
@@ -1,174 +1,194 @@
   class(luplot$plotstyle) <- "data.frame"
-  
+
   # choose plot colors for given entities
   entity <- c(...)
-  if(is.null(entity)) {
+  if (is.null(entity)) {
     entity <- row.names(luplot$plotstyle)
   } else {
     entity[is.na(entity)] <- "NA"
-    entity <- unlist(lapply(strsplit(entity," \\("),function(x) x[1]))    
+    entity <- unlist(lapply(strsplit(entity, " \\("), function(x) x[1]))
   }
-  
-  uq_entity <- unique(entity)
-  res <- luplot$plotstyle[uq_entity,]
-  row.names(res) <- uq_entity
-  
-  # ignore rest of function of regexp is used
+
+  uqEntity <- unique(entity)
+  res <- luplot$plotstyle[uqEntity, ]
+  row.names(res) <- uqEntity
+
+  # ignore rest of function if regexp is used
   if (regexp) {
     res <- data.frame()
-    for (r in uq_entity) {
+    for (r in uqEntity) {
       # get rows that match the regular expression
       indices <- which(grepl(r, row.names(luplot$plotstyle)))
       # add rows found
       if (0 < length(indices)) {
-        new.rows <- luplot$plotstyle[indices,]
+        newRows <- luplot$plotstyle[indices, ]
       } else {
         # if none are found, add an 'NA'-row
-        new.rows <- luplot$plotstyle[paste(row.names(luplot$plotstyle), 
-                                           collapse = ''),]
+        newRows <- luplot$plotstyle[paste(row.names(luplot$plotstyle),
+                                           collapse = ""), ]
         # with the regular expression as name
-        row.names(new.rows) <- r
+        row.names(newRows) <- r
       }
-      res <- rbind(res, new.rows) 
+      res <- rbind(res, newRows)
     }
-    
-    if ('all' == out) {
+
+    if ("all" == out) {
       return(res)
     } else {
       return(setNames(getElement(res, out), rownames(res)))
     }
   }
-  
-  # count unknown entities, i.e. count rows that have NA only, i.e. where number of columns is the same as number of NAs
-  ina <- rowSums(is.na(res))==NCOL(res)
+
+  # count unknown entities, i.e. count rows that have NA only,
+  # i.e. where number of columns is the same as number of NAs
+  ina <- rowSums(is.na(res)) == NCOL(res)
   nna <- sum(ina)
 
   # replace NA
   if (nna != 0) {
     if (is.null(unknown)) {
-      # if you need less than 10 colors brew and ramp for the exact number of colors min(9,nna) -> nna (ramping has no effect -> take the original brewer colors)
-      # only if you need more than 9 colors start ramping between maximal 9 brewer colors effectively
-      # replace NA in color with freshly brewed colors from "Set1"
-      tmpcols <- brewer.pal(9,"Set1")
-      tmpcols[6] <- "#00CED1" #replace yellow with turquoise as it was to light for many applications
-      if(nna<9) tmpcols <- tmpcols[1:nna]
-      set1 <- colorRampPalette(tmpcols)
+      # The following list contains 19 easily distinguishable colors. If you
+      # need n < 20 colors, choose the first n colors from that list. If you
+      # need more colors, colorRampPalette() is used to create interpolated
```

Comment:
still, if you have 20 or more colors, the first 19 are not stable, but the question is whether this bothers someone. With 20 lines, you probably have other problems :)

An alternative could be to define `colorPalette` as:
``` R
colorPalette <- function(n) {
  return(c(tmpcols[1:min(n, length(tmpcols))], heat.colors(max(0, n-length(tmpcols)))))
}
```
This way, the first 19 colors are never changed and simply further colors are added, and only those are not stable.

Reply:
Good idea! Inspired by you comment, I change it so that it is always consistent. But maybe this was overkill... 😁

Reply:
I can confirm that it is certainly overkill. A pity that you made it deterministic :)